src/Lib/package.cpp

```diff
@@ -34,17 +40,61 @@ const std::string Package::GetFullyQualifiedName(void) {
   return Package::GetName() + " " + Package::GetVersion();
 }
 
+const void Package::assignDataDir(void) {
+  const char *possible[] = {".", "..", DATADIR, NULL};
+  int i = 0;
+
+  while (possible[i] != NULL) {
+    char map[1024] = "";
+    size_t r = snprintf(map, sizeof map, "%s/maps", possible[i]);
+    if (r >= sizeof map) {
+      // handle error
+    }
+
+    std::ifstream file(map);
+    if (file.good()) {
+      std::string leadPath(possible[i]);
+      Package::setDataDir(leadPath);
+      return;
+    }
+
+    i++;
+  }
+
+  // check env for $APPDIOR
+
+  return;
+}
+
 #else
 
+#include <filesystem>
+
 #include "package.hpp"
 #include "test.hpp"
 
-int main(void)
-{
+void test_name(void) {
   int len = Package::GetVersion().length();
   assert(len >= 5 && len < 10);
   assert(Package::GetName().compare("NetPanzer") == 0);
 
+  return;
+}
+
+void test_datadir(void) {
+  // fprintf(stderr, "meson_build_root: " MESON_BUILD_ROOT "\n");
+  std::filesystem::current_path(std::string(MESON_BUILD_ROOT));
+  // fprintf(stderr, "cwd: %s\n", path.c_str());
+  Package::assignDataDir();
+  // fprintf(stdout, "datadir: %s\n", Package::getDataDir().c_str());
+  assert(Package::getDataDir() == "..");
```

Comment:
hm, so dataDir is relative to the binary path here, but when someone passes DATA_DIR won't they want to pass an absolute path? will the code break/be confusing when dealing with both abs and relative paths?

Reply:
I'm not sure yet. DATADIR will be hardcoded at compile time, mostly by distro maintainers. Actually, that's why DATADIR should be checked for last. If we're testing, we don't want NP to look at the hard-coded DATADIR. Or maybe there's a way not to define it in config.h... whatever. I'm tired now. lol

Reply:
Ah I misread, I was thinking DATADIR was the env variable for some reason. So yeah, we have to check the env variable first, I guess.

> Actually, that's why DATADIR should be checked for last

It's an interesting question. I guess what matters is, will it pickup the right path. When the distro maintainer sets it, the binary probably won't be in a location that contains the `maps` path anyway, so it probably doesn't matter if it's first or last?

Reply:
When packaged by maintainers, DATADIR will be set to an absolute path.

For development, DATADIR will be set to some absolute path like /usr/local/share/netpanzer but it will be ignored in favor of where the development data is.

I'll look into more what Eli said at https://github.com/netpanzer/netpanzer/pull/46#issuecomment-1877629802

Right now the binary is still only working when it's in the source root.

I changed it so it uses absolute paths but it made no difference.